src/Microsoft.Azure.SignalR.Common/ServiceConnections/ServiceConnectionContainerBase.cs

```diff
@@ -301,6 +309,39 @@ protected virtual ServiceConnectionStatus GetStatus()
             return FixedServiceConnections.Any(s => s.Status == ServiceConnectionStatus.Connected)
                 ? ServiceConnectionStatus.Connected
                 : ServiceConnectionStatus.Disconnected;
+        }
+
+        protected async Task WriteFinAsync(IServiceConnection c, bool migratable)
+        {
+            if (migratable)
+            {
+                await c.WriteAsync(_shutdownFinMigratableMessage);
+            }
+            else
+            {
+                await c.WriteAsync(_shutdownFinMessage);
+            }
+        }
+
+        protected async Task RemoveConnectionAsync(IServiceConnection c, bool migratable)
+        {
+            var retry = 0;
+            using var source = new CancellationTokenSource();

```

Comment:
move init into the loop, and add some test?

Reply:
Fixed cancellation token source. And leave migration related tests for @terencefan. live-scale related tests will be added after merged most PRs for complete e2e tests.